### PR TITLE
Fix/nex 1641/assets compilation

### PR DESCRIPTION
**Pull request by boajer**
FIX: TAO Community - Cannot Compile Deliveries with Remote Assets
Jira: https://oat-sa.atlassian.net/browse/NEX-1641

Related PR
- [ ] https://github.com/oat-sa/extension-tao-item/pull/424

### CODE REVIEW

File: model/QtiJsonItemCompiler.php [hunk 1]
```diff
@@ -100,10 +100,9 @@ protected function deployQtiItem(
 
 
         try {
-
             $qtiItem = $this->createQtiItem($item, $language);
-            $resolver =  $resolver = new ItemMediaResolver($item, $language);
-            $publicLangDirectory = $publicDirectory->getDirectory($language);;
+            $resolver = $resolver = new ItemMediaResolver($item, $language);
+            $publicLangDirectory = $publicDirectory->getDirectory($language);
```
Comment (mike-ionut-mihai-sandu-tao) [inline model/QtiJsonItemCompiler.php:105]: nice

File: model/compile/QtiAssetCompiler/QtiItemAssetCompiler.php [hunk 2]
```diff
@@ -87,10 +86,14 @@ private function resolve(ItemMediaResolver $resolver, string $assetUrl, string $
         $mediaAsset = $resolver->resolve($assetUrl);
         $mediaSource = $mediaAsset->getMediaSource();
 
+        if ($mediaSource instanceof HttpSource) {
```
Comment (mike-ionut-mihai-sandu-tao) [inline model/compile/QtiAssetCompiler/QtiItemAssetCompiler.php:89]: looks good

File: model/compile/QtiAssetCompiler/QtiItemAssetCompiler.php [hunk 3]
```diff
@@ -101,20 +104,7 @@ private function resolve(ItemMediaResolver $resolver, string $assetUrl, string $
...
-        while (in_array($replacement, $replacementList)) {
-            $dot = strrpos($basename, '.');
-            $replacement = $dot !== false
-                ? substr($basename, 0, $dot) . '_' . $count . substr($basename, $dot)
-                : $basename . $count;
-            $count++;
-        }
-
-        return $replacement;
+        return hash('crc32', basename($link));
```
Comment (mike-ionut-mihai-sandu-tao) [inline model/compile/QtiAssetCompiler/QtiItemAssetCompiler.php:107]: may i ask you why you want to use `crc32` and not another algorithm

**Review by mike-ionut-mihai-sandu-tao — APPROVED** [review]
lgtm

File: model/compile/QtiAssetCompiler/QtiItemAssetCompiler.php [hunk 4]
```diff
@@ -101,20 +104,7 @@ private function resolve(ItemMediaResolver $resolver, string $assetUrl, string $
...
-
-        $count = 0;
-        while (in_array($replacement, $replacementList)) {
-            $dot = strrpos($basename, '.');
-            $replacement = $dot !== false
-                ? substr($basename, 0, $dot) . '_' . $count . substr($basename, $dot)
-                : $basename . $count;
-            $count++;
-        }
-
```
Comment (bugalot) [inline model/compile/QtiAssetCompiler/QtiItemAssetCompiler.php:116]: Is this the only real change?

Comment (boajer) [reply model/compile/QtiAssetCompiler/QtiItemAssetCompiler.php:116]: Nope, as discussed in slack https://github.com/oat-sa/extension-tao-itemqti/pull/1517/files#diff-e8652df0cba6546296d14104113388f6R89 this is required too

Comment (mike-ionut-mihai-sandu-tao) [reply model/compile/QtiAssetCompiler/QtiItemAssetCompiler.php:116]: also : https://github.com/oat-sa/extension-tao-itemqti/pull/1517/files#diff-e8652df0cba6546296d14104113388f6R89

File: model/compile/QtiAssetCompiler/XIncludeXmlInjector.php [hunk 5]
```diff
@@ -38,18 +38,22 @@ class XIncludeXmlInjector extends ConfigurationService
...
+        $original = $domDocument->getElementsByTagName('include');
+        $elements = [];
+
         /** @var DOMElement $xincludeNode */
-        foreach ($domDocument->getElementsByTagName('include') as $xincludeNode) {
+        foreach ($original as $xincludeNode) {
+            $elements[] = $xincludeNode;
+        }
 
+        foreach ($elements as $xincludeNode) {
```
Comment (gabrielfs7) [inline model/compile/QtiAssetCompiler/XIncludeXmlInjector.php:51]: Why this change? It is doing same as before, no?

File: model/compile/QtiAssetCompiler/QtiItemAssetCompiler.php (hunk 3 above)
Comment (gabrielfs7) [reply model/compile/QtiAssetCompiler/QtiItemAssetCompiler.php:107]: Wont this break some `decode` made in other place?

File: model/compile/QtiAssetCompiler/XIncludeXmlInjector.php (hunk 5 above)
Comment (boajer) [reply model/compile/QtiAssetCompiler/XIncludeXmlInjector.php:51]: before: 
 `replaceChild` called on the iterable nodes collection made changes to it; That lead to fact internal pointer pointed to not correct element of collection and not all of them replaced in reality. Literally it was skipping even shared stimulues. 
after:
Copiying to simple array allowes to use simple iterations and avoide loosing elements

You can visualize it with an easy example - try to run delivery with 1 item, containing 5 shared stimuluses ( no matter withing 1 interaction or each of them can be wrapped into own )

Comment (gabrielfs7) [reply model/compile/QtiAssetCompiler/XIncludeXmlInjector.php:51]: ok, I got. We were updating the same object inside the loop, thanks!

File: model/compile/QtiAssetCompiler/QtiItemAssetCompiler.php [hunk 6]
```diff
@@ -101,20 +104,7 @@ private function resolve(ItemMediaResolver $resolver, string $assetUrl, string $
 
     private function getReplacementName(string $link, array $replacementList): string
```
Comment (gabrielfs7) [inline model/compile/QtiAssetCompiler/QtiItemAssetCompiler.php:105]:
```suggestion
    private function getReplacementName(string $link): string
```
Parameter can be removed now

File: model/compile/QtiAssetCompiler/QtiItemAssetCompiler.php [hunk 7]
```diff
@@ -50,9 +51,7 @@ public function extractAndCopyAssetFiles(Item $qtiItem, Directory $publicDirecto
         $replacementList = $packedAssets = [];
```
Comment (gabrielfs7) [inline model/compile/QtiAssetCompiler/QtiItemAssetCompiler.php:51]: This is no longer needed. Also the calls must be fixed: https://github.com/oat-sa/extension-tao-itemqti/pull/1517/files#diff-e8652df0cba6546296d14104113388f6R61

File: model/compile/QtiAssetCompiler/QtiItemAssetCompiler.php (hunk 6 above)
Comment (mike-ionut-mihai-sandu-tao) [reply model/compile/QtiAssetCompiler/QtiItemAssetCompiler.php:105]: also if you do this change please remove line : 53, 64 and 67

File: model/compile/QtiAssetCompiler/XIncludeXmlInjector.php [hunk 8]
```diff
@@ -38,18 +38,22 @@ class XIncludeXmlInjector extends ConfigurationService
      * @param DOMDocument $domDocument
      * @param PackedAsset[] $packedAssets
      */
-    public function injectSharedStimulus(DOMDocument $domDocument, array $packedAssets) : void
+    public function injectSharedStimulus(DOMDocument $domDocument, array $packedAssets): void
     {
+        $original = $domDocument->getElementsByTagName('include');
```
Comment (mike-ionut-mihai-sandu-tao) [inline model/compile/QtiAssetCompiler/XIncludeXmlInjector.php:43]: a bit misleading the name ` $original`

**Comment by gabrielfs7** [comment]
- [ ] New code is covered by tests (if applicable)
- [ ] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [ ] New code is respecting code style rules
- [ ] New code is respecting best practices
- [ ] New code is not subject to concurrency issues (if applicable)
- [ ] Feature is working correctly on my local machine (if applicable)
- [ ] Acceptance criteria are respected
- [ ] Pull request title and description are meaningful

**Review by gabrielfs7 — APPROVED** [review]
Approved based on Demo, but would like to see input from @bugalot

File: model/compile/QtiAssetCompiler/QtiItemAssetCompiler.php [hunk 9]
```diff
@@ -47,19 +48,17 @@ public function extractAndCopyAssetFiles(Item $qtiItem, Directory $publicDirecto
...
                     continue;
                 }
 
                 $packedAsset = $this->resolve($resolver, $assetUrl, $type);
 
-                $replacement = $this->getReplacementName($packedAsset->getLink(), $replacementList);
+                $replacement = $this->getReplacementName($packedAsset->getLink());
                 $packedAsset->setReplacedBy($replacement);
 
                 $replacementList[$assetUrl] = $replacement;
```
Comment (gabrielfs7) [inline model/compile/QtiAssetCompiler/QtiItemAssetCompiler.php:64]: Still missing this line to be removed

**Review by siwane — APPROVED** [review]
All test cases tested:
test-with-assets

- [ ] New code is covered by tests (if applicable)
- [ ] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [ ] New code is respecting code style rules
- [ ] New code is respecting best practices
- [ ] New code is not subject to concurrency issues (if applicable)
- [ ] Feature is working correctly on my local machine (if applicable)
- [ ] Acceptance criteria are respected
- [ ] Pull request title and description are meaningful
